velox/functions/sparksql/ArrayInsertFunction.h

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/functions/Udf.h"
+
+namespace facebook::velox::functions::sparksql {
+
+/// array_insert(array(E), pos, E, bool) → array(E)
+/// Places new element into index pos of the input array.
+template <typename T>
+struct ArrayInsertFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+
+  static constexpr int32_t kMaxNumberOfElements = 10'000;
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Array<Generic<T1>>>& out,
+      const arg_type<Array<Generic<T1>>>* srcArray,
+      const arg_type<int32_t>* pos,
+      const arg_type<Generic<T1>>* item,
+      const arg_type<bool>* legacyNegativeIndex) {
+    VELOX_USER_CHECK_NOT_NULL(
+        legacyNegativeIndex,
+        "Constant legacyNegativeIndex should be provided.");
```

Comment:
This check only ensures legacyNegativeIndex is not null. Shall we check if it is constant and get its value in an initialize method?

Reply:
>  Shall we check if it is constant and get its value in an initialize method?

This can be guaranteed from spark side. I think the check might be not required here. Do you have any concerns about this?

Reply:
I think this check is needed as user may send non-const inputs and we need to prevent that. But we don't need this check for each row, instead we could use an initialize method to make sure they are constants, seeing https://github.com/facebookincubator/velox/blob/main/velox/docs/develop/scalar-functions.rst#access-to-session-properties-and-constant-inputs.

Reply:
Sounds good. Updated, thanks.